rollup/internal/controller/relayer/full_recovery.go

```diff
@@ -0,0 +1,356 @@
+package relayer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core"
+	"github.com/scroll-tech/go-ethereum/ethclient"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rollup/l1"
+	"gorm.io/gorm"
+
+	"scroll-tech/common/types"
+	"scroll-tech/rollup/internal/config"
+	"scroll-tech/rollup/internal/controller/watcher"
+	"scroll-tech/rollup/internal/orm"
+	butils "scroll-tech/rollup/internal/utils"
+)
+
+type FullRecovery struct {
+	ctx       context.Context
+	cfg       *config.Config
+	genesis   *core.Genesis
+	db        *gorm.DB
+	blockORM  *orm.L2Block
+	chunkORM  *orm.Chunk
+	batchORM  *orm.Batch
+	bundleORM *orm.Bundle
+
+	chunkProposer  *watcher.ChunkProposer
+	batchProposer  *watcher.BatchProposer
+	bundleProposer *watcher.BundleProposer
+	l2Watcher      *watcher.L2WatcherClient
+	l1Client       *ethclient.Client
+	l1Reader       *l1.Reader
+}
+
+func NewFullRecovery(ctx context.Context, cfg *config.Config, genesis *core.Genesis, db *gorm.DB, chunkProposer *watcher.ChunkProposer, batchProposer *watcher.BatchProposer, bundleProposer *watcher.BundleProposer, l2Watcher *watcher.L2WatcherClient, l1Client *ethclient.Client, l1Reader *l1.Reader) *FullRecovery {
+	return &FullRecovery{
+		ctx:       ctx,
+		cfg:       cfg,
+		genesis:   genesis,
+		db:        db,
+		blockORM:  orm.NewL2Block(db),
+		chunkORM:  orm.NewChunk(db),
+		batchORM:  orm.NewBatch(db),
+		bundleORM: orm.NewBundle(db),
+
+		chunkProposer:  chunkProposer,
+		batchProposer:  batchProposer,
+		bundleProposer: bundleProposer,
+		l2Watcher:      l2Watcher,
+		l1Client:       l1Client,
+		l1Reader:       l1Reader,
+	}
+}
+
+// RestoreFullPreviousState restores the full state from L1.
+// The DB state should be clean: the latest batch in the DB should be finalized on L1. This function will
+// restore all batches between the latest finalized batch in the DB and the latest finalized batch on L1.
+func (f *FullRecovery) RestoreFullPreviousState() error {
```

Comment:
Why I feel we can combine the `RestoreMinimalPreviousState` and `RestoreFullPreviousState` to same logic. WDYT？

Reply:
I agree, there is some overlap but if you look at it side by side, the functionality diverges already at 1.5 steps. As such, I think it would make it more complicated to combine them.
<img width="2208" alt="image" src="https://github.com/user-attachments/assets/50b32678-31be-446c-ac9e-7a27d2b33aa2">